src/goto-programs/restrict_function_pointers.h

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************\
+
+Module: GOTO Program Utilities
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+/// \file
+/// Given goto functions and a list of function parameters or globals
+/// that are function pointers with lists of possible candidates, replace use of
+/// these function pointers with calls to the candidate.
+/// The purpose here is to avoid unnecessary branching
+/// i.e. "there are 600 functions with this signature, but I know it's
+/// always going to be one of these two"
+
+#ifndef CPROVER_GOTO_PROGRAMS_RESTRICT_FUNCTION_POINTERS_H
+#define CPROVER_GOTO_PROGRAMS_RESTRICT_FUNCTION_POINTERS_H
+
+#include <unordered_map>
+#include <unordered_set>
+
+#include <util/cmdline.h>
+#include <util/irep.h>
+
+#include <goto-programs/goto_model.h>
+#include <util/options.h>
+
+#define RESTRICT_FUNCTION_POINTER_OPT "restrict-function-pointer"
+#define RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT                                \
+  "function-pointer-restrictions-file"
+
+#define OPT_RESTRICT_FUNCTION_POINTER                                          \
+  "(" RESTRICT_FUNCTION_POINTER_OPT                                            \
+  "):"                                                                         \
+  "(" RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT "):"
+
+#define RESTRICT_FUNCTION_POINTER_HELP                                         \
+  "--" RESTRICT_FUNCTION_POINTER_OPT                                           \
+  " <pointer_name>/<target[,targets]*>\n"                                      \
+  "           restrict a function pointer to a set of possible targets\n"      \
```

Comment:
This would definitely benefit with some examples in some documentation somewhere

Reply:
I believe this is explained in the cprover-manual